R/tax_range_geo.R

```diff
@@ -0,0 +1,268 @@
+    row.names(spat_df) <- NULL
+    return(spat_df)
+  }
+
+  #=== Latitudinal range ===
+  if (method == "lat") {
+    # Generate dataframe for population
+    lat_df <- data.frame(taxa = unique_taxa,
+                         taxon_id = seq(1, length(unique_taxa), 1),
+                         max_lat = rep(NA, length(unique_taxa)),
+                         min_lat = rep(NA, length(unique_taxa)),
+                         range_lat = rep(NA, length(unique_taxa)))
+    # Run for loop across unique taxa
+    for(i in seq_along(unique_taxa)){
+      vec <- which(occdf[, name] == unique_taxa[i])
+      lat_df$max_lat[i] <- max(occdf[vec, lat])
+      lat_df$min_lat[i] <- min(occdf[vec, lat])
+      lat_df$range_lat[i] <- lat_df$max_lat[i] - lat_df$min_lat[i]
+    }
+    # Remove row names
+    row.names(lat_df) <- NULL
+
+    # Round off values
+    lat_df[, c("max_lat", "min_lat", "range_lat")] <- round(
+      x = lat_df[, c("max_lat", "min_lat", "range_lat")], digits = 3)
+
+    # Return dataframe
+    return(lat_df)
+  }
+
+  #=== Great Circle Distance  ===
+  if (method == "gcd") {
+    # Generate dataframe for population
+    gcd_df <- data.frame()
+    # Run for loop across unique taxa
+    for(i in seq_along(unique_taxa)){
+      # Unique taxa name
+      taxa <- unique_taxa[i]
+      # taxon id
+      taxon_id <- i
+      # Subset df
+      tmp <- occdf[which(occdf[, name] == unique_taxa[i]), ]
+      # Calculate GCD matrix using the Haversine method
+      vals <- geosphere::distm(x = tmp[,c(lng, lat)],
+                               fun = geosphere::distHaversine)
+      # Convert to km
+      vals <- vals / 10^3
+      # Extract location of points with max GCD
+      loc <- which(vals == max(vals), arr.ind = TRUE)
+      # Get maximum GCD in km
+      GCD <- as.numeric(max(vals))
+      # Extract coordinates of points
+      coords <- data.frame(tmp[loc[1, 1:2], c(lng, lat)])
+      # Build dataframe
+      tmp <- cbind.data.frame(taxa, taxon_id, coords, GCD)
+      gcd_df <- rbind.data.frame(gcd_df, tmp)
+    }
+
+    # Remove row names
+    row.names(gcd_df) <- NULL
+
+    # Round off values
+    gcd_df[, c(lng, lat, "GCD")] <- round(
+      x = gcd_df[, c(lng, lat, "GCD")], digits = 3)
+
+    # Return dataframe
+    return(gcd_df)
+  }
+  #=== Occupied grid cells  ===
+  if (method == "occ") {
+    # Generate dataframe for population
+    oc_df <- data.frame(taxa = unique_taxa,
+                         taxon_id = seq(1, length(unique_taxa), 1),
+                         cells = rep(NA, length(unique_taxa)),
+                         spacing = rep(NA, length(unique_taxa)))
+    # Generate equal area hexagonal grid
+    # Which resolution should be used based on input distance/spacing?
+    # Use the h3jsr::h3_info_table to calculate resolution
+    grid <- h3jsr::h3_info_table[
+      which.min(abs(h3jsr::h3_info_table$avg_cendist_km - spacing)), ]
+    # Add resolution spacing
+    oc_df$spacing <- grid$avg_cendist_km
```

Comment:
Would it be better to return 'spacing' as a separate object instead of adding it to this dataframe? It seems a bit redundant to just have the same information listed over and over again (I guess it could be useful for further analysis as a category, but it's a bit of a stretch). 

One option would be to return a list, with the dataframe as one entry and the spacing as another. Unsure if this is something that you'd want to implement as it does make getting the dataframe marginally harder, but it would make it a lot more elegant. Just a thought!

Reply:
As the number/proportion of occupied cells is hugely dependent on this value, I think it is good for the user to have a column with this data. An alternative would be to have a message returned but this would just get lost if the user is not actively watching the console (i.e. pipeline analysis). I see this as comparable as having a column in your dataframe for quorum level in SQS... ultimately, much easier for a user to drop a column, than to add it manually or handle a list?

Reply:
Yeah, I think now you've added prop. of occupied cells this should be kept. You could have a case where you want to vary the size of grids, and test proportions - in which case you'd want that kept in long format. Happy with this!